controllers/gateway/reconcile_result.go

```diff
@@ -15,7 +15,7 @@ import (
 )
 
 // Updates api status. If there was an error during update, returns the error so that entire reconcile loop is retried. If there is no error, returns a "reconcile success" value.
-func (r *APIRuleReconciler) updateStatusOrRetry(ctx context.Context, api *gatewayv1beta1.APIRule, status status.ReconciliationStatus) (ctrl.Result, error) {
+func (r *APIRuleReconciler) updateStatusOrRetry(ctx context.Context, api *gatewayv1beta1.APIRule, status status.ReconciliationStatus, needsMigration ...bool) (ctrl.Result, error) {
```

Comment:
I personally would copy & paste some of the code and have a dedicated function that is invoked in case we need a migration. This way we have a clear separation between the logic and it's also easier to remove the migration logic later on.